File: backend/src/worker.js
import { Worker } from 'bullmq';
import { chromium } from 'playwright';
import { Readability } from '@mozilla/readability';
import { JSDOM } from 'jsdom';
import crypto from 'crypto';
import { connection } from './queue.js';
import { query } from './db.js';
import { getEmbedding } from './embeddings.js';
import { encryptBuffer } from './crypto.js';
import { uploadAsset } from './s3.js';
import { checkOllamaStatus, generateSummary, generateTags } from './ollama.js';

let browser = null;

// Initialize Playwright Browser
const getBrowser = async () => {
  if (!browser) {
    console.log('Launching headless Playwright Chromium instance...');
    browser = await chromium.launch({
      headless: true,
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-web-security'
      ]
    });
  }
  return browser;
};

// Predict Category Name via LLM or fallback
const predictCategory = async (text, title, url) => {
  const isOnline = await checkOllamaStatus();
  const ollamaUrl = process.env.OLLAMA_URL || 'http://localhost:11434';
  const ollamaModel = process.env.OLLAMA_MODEL || 'llama3';

  if (isOnline) {
    try {
      const prompt = `You are a categorization assistant. Given the title, URL, and content snippet of a web page, predict a single, broad category name (e.g., "Technology", "Cooking", "Finance", "Science", "Education", "Lifestyle", "Design", "News").
Respond ONLY with the category name (1-2 words). Do not include any punctuation, quotes, or extra text.
Title: ${title}
URL: ${url}
Snippet: ${text.slice(0, 1000)}

Category:`;

      const response = await fetch(`${ollamaUrl}/api/generate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          model: ollamaModel,
          prompt,
          stream: false
        })
      });

      if (response.ok) {
        const json = await response.json();
        const categoryName = json.response.trim().replace(/[^a-zA-Z0-9\s/]/g, '');
        if (categoryName && categoryName.length < 30) {
          return categoryName;
        }
      }
    } catch (err) {
      console.warn('Ollama category prediction failed, using fallback.');
    }
  }

  // Fallback Rule-Based Categorization
  const lowerText = `${title} ${url} ${text.slice(0, 1000)}`.toLowerCase();
  
  if (lowerText.includes('github') || lowerText.includes('code') || lowerText.includes('npm') || lowerText.includes('developer') || lowerText.includes('react') || lowerText.includes('api')) {
    return 'Development';
  }
  if (lowerText.includes('recipe') || lowerText.includes('cook') || lowerText.includes('food') || lowerText.includes('kitchen') || lowerText.includes('bake')) {
    return 'Cooking';
  }
  if (lowerText.includes('stock') || lowerText.includes('crypto') || lowerText.includes('finance') || lowerText.includes('investing') || lowerText.includes('money')) {
    return 'Finance';
  }
  if (lowerText.includes('design') || lowerText.includes('css') || lowerText.includes('ui/ux') || lowerText.includes('vector') || lowerText.includes('color')) {
    return 'Design';
  }
  if (lowerText.includes('science') || lowerText.includes('physics') || lowerText.includes('research') || lowerText.includes('nature') || lowerText.includes('space')) {
    return 'Science';
  }

  return 'General';
};

// Process Bookmark Ingestion Job
const processIngestion = async (job) => {
  const { bookmarkId, userId, url, encryptionKeyHex } = job.data;
  console.log(`[Job ${job.id}] Starting ingestion for Bookmark ${bookmarkId} (URL: ${url})`);

  let pageContext;
  try {
    const playwrightBrowser = await getBrowser();
    pageContext = await playwrightBrowser.newContext({
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      viewport: { width: 1280, height: 800 }
    });

    const page = await pageContext.newPage();
    
    // Set 30s timeout
    await page.goto(url, { waitUntil: 'load', timeout: 30000 });
    
    // Wait an extra second for layout
    await page.waitForTimeout(1000);

    const title = (await page.title()) || new URL(url).hostname;
    const htmlDump = await page.content();
    
    // Capture screenshot (PNG)
    const screenshotBuffer = await page.screenshot({ type: 'png', fullPage: false });
    
    // Capture PDF (Print format) - wrapped in try-catch as PDF printing is only supported in headless Chromium
    let pdfBuffer = Buffer.alloc(0);
    try {
      pdfBuffer = await page.pdf({ format: 'A4', printBackground: true });
    } catch (pdfErr) {
      console.warn('PDF printing not supported or failed:', pdfErr.message);
    }

    // Extract page metadata and main text with Readability
    const dom = new JSDOM(htmlDump, { url });
    const doc = dom.window.document;
    const metaDescription = doc.querySelector('meta[name="description"]')?.getAttribute('content') || '';
    
    const reader = new Readability(doc);
    const parsedArticle = reader.parse();
    const readableText = parsedArticle ? parsedArticle.textContent : doc.body.innerText || '';

    // Generate local vector embedding (384 Dimensions)
    const textEmbedding = await getEmbedding(readableText || title);

    // Retrieve categories and match centroid using pgvector cosine distance
    const catMatchRes = await query(
      `SELECT id, name, (centroid_vector <=> $1::vector) AS distance 
       FROM categories 
       WHERE user_id = $2 AND centroid_vector IS NOT NULL 
       ORDER BY distance ASC 
       LIMIT 1`,
      [JSON.stringify(textEmbedding), userId]
    );

    let finalCategoryId = null;

    if (catMatchRes.rows.length > 0 && catMatchRes.rows[0].distance < 0.25) {
      // High-confidence match: assign to existing category
      finalCategoryId = catMatchRes.rows[0].id;
      console.log(`Auto-assigned to existing category: "${catMatchRes.rows[0].name}" (Distance: ${catMatchRes.rows[0].distance})`);
    } else {
      // No close centroid match. Let's predict a category
      const categoryName = await predictCategory(readableText, title, url);
      
      // Check if user already has a category with this name
      const checkCatRes = await query(
        'SELECT id FROM categories WHERE user_id = $1 AND LOWER(name) = LOWER($2)',
        [userId, categoryName]
      );

      if (checkCatRes.rows.length > 0) {
        finalCategoryId = checkCatRes.rows[0].id;
        console.log(`Matched predicted category with existing: "${categoryName}"`);
      } else {
        // Create new category
        const createCatRes = await query(
          'INSERT INTO categories (user_id, name, description) VALUES ($1, $2, $3) RETURNING id',
          [userId, categoryName, `Auto-generated category for ${categoryName}-related content.`]
        );
        finalCategoryId = createCatRes.rows[0].id;
        console.log(`Created new category: "${categoryName}"`);
      }
    }

    // Generate summary and tags via Ollama or fallbacks
    const summary = await generateSummary(readableText, title, metaDescription);
    const tags = await generateTags(readableText, title);

    // Update bookmark metadata & vector
    await query(
      `UPDATE bookmarks 
       SET title = $1, summary = $2, category_id = $3, raw_text_vector = $4::vector 
       WHERE id = $5`,
      [title, summary, finalCategoryId, JSON.stringify(textEmbedding), bookmarkId]
    );

    // Associate Tags
    for (const tagName of tags) {
      // Insert tag if it doesn't exist
      const tagInsert = await query(
        `INSERT INTO tags (user_id, name) 
         VALUES ($1, $2) 
         ON CONFLICT (user_id, name) DO UPDATE SET name = EXCLUDED.name 
         RETURNING id`,
        [userId, tagName]
      );
      const tagId = tagInsert.rows[0].id;

      // Associate with bookmark
      await query(
        `INSERT INTO bookmark_tags (bookmark_id, tag_id) 
         VALUES ($1, $2) 
         ON CONFLICT DO NOTHING`,
        [bookmarkId, tagId]
      );
    }

    // Encrypt and Upload Assets if encryption key is supplied
    if (encryptionKeyHex) {
      const assets = [
        { type: 'html_dump', buffer: Buffer.from(htmlDump), mime: 'text/html' },
        { type: 'screenshot', buffer: screenshotBuffer, mime: 'image/png' }
      ];

      if (pdfBuffer && pdfBuffer.length > 0) {
        assets.push({ type: 'pdf', buffer: pdfBuffer, mime: 'application/pdf' });
      }

      for (const asset of assets) {
        // Perform server-side symmetric encryption
        const { encrypted, ivHex } = encryptBuffer(asset.buffer, encryptionKeyHex);
        const sha256 = crypto.createHash('sha256').update(asset.buffer).digest('hex');

        // Path structure: user_id/bookmark_id/asset_type
        const storagePath = `${userId}/${bookmarkId}/${asset.type}.enc`;

        // Upload to S3/MinIO
        await uploadAsset(storagePath, encrypted, asset.mime);

        // Store encrypted asset record in DB
        await query(
          `INSERT INTO encrypted_assets (bookmark_id, asset_type, storage_path, initialization_vector, sha256_checksum)
           VALUES ($1, $2, $3, $4, $5)`,
          [bookmarkId, asset.type, storagePath, ivHex, sha256]
        );
      }
      console.log(`Scraped and encrypted all assets for Bookmark ${bookmarkId}`);
    } else {
      console.warn(`No encryption key provided for Bookmark ${bookmarkId}. Assets skipped.`);
    }

    console.log(`[Job ${job.id}] Finished processing successfully.`);
  } catch (error) {
    console.error(`[Job ${job.id}] Failed to process bookmark ingestion:`, error);
    
    // Safe fallback: Update bookmark status to let user know scraping failed
    await query(
      `UPDATE bookmarks 
       SET title = COALESCE(title, $1), summary = $2 
       WHERE id = $3`,
      [new URL(url).hostname, `Failed to scrape page content. Error: ${error.message}`, bookmarkId]
    );
    throw error;
  } finally {
    if (pageContext) {
      await pageContext.close();
    }
  }
};

// Start BullMQ Worker
export const startWorker = () => {
  const worker = new Worker('ingestionQueue', processIngestion, {
    connection,
    concurrency: 2 // run up to 2 scraper jobs in parallel
  });

  worker.on('completed', (job) => {
    console.log(`[Worker] Job ${job.id} has completed!`);
  });

  worker.on('failed', (job, err) => {
    console.error(`[Worker] Job ${job?.id} failed with error: ${err.message}`);
  });

  console.log('BullMQ Ingestion Worker started successfully.');
  return worker;
};

// If worker script is run directly, execute startWorker
if (process.argv[1] && process.argv[1].endsWith('worker.js')) {
  startWorker();
}

File: backend/src/worker/worker.go
package worker

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"

	"linkhub/backend/src/crypto"
	"linkhub/backend/src/db"
	"linkhub/backend/src/s3"

	"github.com/chromedp/cdproto/page"
	"github.com/chromedp/chromedp"
	"github.com/go-redis/go-redis/v9"
	"github.com/go-shiori/go-readability"
)

type IngestJob struct {
	BookmarkID       int    `json:"bookmarkId"`
	UserID           int    `json:"userId"`
	URL              string `json:"url"`
	EncryptionKeyHex string `json:"encryptionKeyHex"`
}

type EmbedResponse struct {
	Embedding []float32 `json:"embedding"`
}

type GenerateResponse struct {
	Response string `json:"response"`
}

var RedisClient *redis.Client

// InitWorker establishes connection to Redis
func InitWorker() {
	redisAddr := os.Getenv("REDIS_URL")
	if redisAddr == "" {
		redisAddr = "localhost:6379"
	}
	// Parse redis:// proto if present
	if strings.HasPrefix(redisAddr, "redis://") {
		redisAddr = strings.TrimPrefix(redisAddr, "redis://")
	}

	RedisClient = redis.NewClient(&redis.Options{
		Addr: redisAddr,
	})
	log.Printf("Connected to Redis at %s for Ingestion Worker.\n", redisAddr)
}

// StartWorker spins up the background queue loop
func StartWorker() {
	ctx := context.Background()
	log.Println("Ingestion Worker loop started.")

	for {
		// BRPOP blocks until a job is available in the queue
		res, err := RedisClient.BRPop(ctx, 0, "linkhub_ingest_queue").Result()
		if err != nil {
			log.Printf("Redis queue BRPOP error: %v\n", err)
			time.Sleep(2 * time.Second)
			continue
		}

		if len(res) < 2 {
			continue
		}

		payload := res[1]
		log.Printf("Processing new ingestion job from queue: %s\n", payload)

		var job IngestJob
		if err := json.Unmarshal([]byte(payload), &job); err != nil {
			log.Printf("Failed to deserialize job payload: %v\n", err)
			continue
		}

		go func(j IngestJob) {
			if err := ProcessJob(j); err != nil {
				log.Printf("Job failed for Bookmark ID %d: %v\n", j.BookmarkID, err)
			} else {
				log.Printf("Job succeeded for Bookmark ID %d\n", j.BookmarkID)
			}
		}(job)
	}
}

// ProcessJob performs scraping, AI extraction, and encrypted uploads
func ProcessJob(job IngestJob) error {
	log.Printf("Scraping URL: %s\n", job.URL)

	// Set up Chromedp
	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.NoSandbox,
		chromedp.DisableGPU,
		chromedp.Flag("disable-setuid-sandbox", true),
		chromedp.Flag("disable-dev-shm-usage", true),
		chromedp.Flag("disable-web-security", true),
	)
	
	allocCtx, allocCancel := chromedp.NewExecAllocator(context.Background(), opts...)
	defer allocCancel()

	ctx, cancel := chromedp.NewContext(allocCtx)
	defer cancel()

	ctx, cancel = context.WithTimeout(ctx, 45*time.Second)
	defer cancel()

	var title string
	var htmlDump string
	var screenshotBuf []byte
	var pdfBuf []byte

	// Execute Chromedp browser actions
	err := chromedp.Run(ctx,
		chromedp.Navigate(job.URL),
		chromedp.Sleep(1500*time.Millisecond),
		chromedp.Title(&title),
		chromedp.OuterHTML("html", &htmlDump),
		chromedp.CaptureScreenshot(&screenshotBuf),
		chromedp.ActionFunc(func(ctx context.Context) error {
			var err error
			// PDF generation only works in headless mode
			pdfBuf, _, err = page.PrintToPDF().WithPrintBackground(true).Do(ctx)
			if err != nil {
				log.Printf("PDF printing skipped or failed: %v\n", err)
			}
			return nil
		}),
	)

	if err != nil {
		fallbackTitle := job.URL
		if u, errParse := url.Parse(job.URL); errParse == nil {
			fallbackTitle = u.Hostname()
		}
		
		// Mark bookmark as failed in DB
		_, dbErr := db.Pool.Exec(context.Background(),
			"UPDATE bookmarks SET title = $1, summary = $2 WHERE id = $3",
			fallbackTitle, fmt.Sprintf("Failed to scrape page. Error: %v", err), job.BookmarkID,
		)
		if dbErr != nil {
			log.Printf("DB error saving failure state: %v\n", dbErr)
		}
		return fmt.Errorf("chromedp run failed: %w", err)
	}

	if title == "" {
		title = "Untitled Bookmark"
	}

	// Parse main text content via go-readability
	parsedUrl, _ := url.Parse(job.URL)
	article, err := readability.FromReader(strings.NewReader(htmlDump), parsedUrl)
	readableText := ""
	if err == nil {
		readableText = article.TextContent
	} else {
		readableText = htmlDump
	}

	// Get AI embeddings
	embedding, err := getOllamaEmbedding(readableText, title)
	if err != nil {
		log.Printf("Ollama embedding extraction failed: %v. Using dummy vector.", err)
		embedding = make([]float32, 384) // 384 dimensions
	}

	// Convert embedding to JSON format for pgvector
	embeddingJson, err := json.Marshal(embedding)
	if err != nil {
		embeddingJson = []byte("[]")
	}

	// Semantic matching for folder assignment
	var finalCategoryId *int
	var matchedCatName string
	var distance float64

	// Query closest category by cosine similarity
	err = db.Pool.QueryRow(context.Background(),
		`SELECT id, name, (centroid_vector <=> $1::vector) AS distance 
		 FROM categories 
		 WHERE user_id = $2 AND centroid_vector IS NOT NULL 
		 ORDER BY distance ASC LIMIT 1`,
		string(embeddingJson), job.UserID,
	).Scan(&finalCategoryId, &matchedCatName, &distance)

	if err == nil && distance < 0.25 {
		log.Printf("Auto-assigned to existing folder: \"%s\" (Distance: %f)\n", matchedCatName, distance)
	} else {
		// Category prediction fallback
		predCategoryName := predictCategoryName(readableText, title, job.URL)
		
		// Check if folder exists
		var existingId int
		err = db.Pool.QueryRow(context.Background(),
			"SELECT id FROM categories WHERE user_id = $1 AND LOWER(name) = LOWER($2)",
			job.UserID, predCategoryName,
		).Scan(&existingId)

		if err == nil {
			finalCategoryId = &existingId
			log.Printf("Matched predicted category with folder: \"%s\"\n", predCategoryName)
		} else {
			// Create new category
			var newId int
			err = db.Pool.QueryRow(context.Background(),
				"INSERT INTO categories (user_id, name, description) VALUES ($1, $2, $3) RETURNING id",
				job.UserID, predCategoryName, fmt.Sprintf("Auto-generated folder for %s-related content.", predCategoryName),
			).Scan(&newId)
			if err == nil {
				finalCategoryId = &newId
				log.Printf("Created new folder: \"%s\"\n", predCategoryName)
			}
		}
	}

	// AI summary & tags
	summary := generateSummary(readableText, title)
	tags := generateTags(readableText, title)

	// Save main Bookmark metadata
	_, err = db.Pool.Exec(context.Background(),
		`UPDATE bookmarks 
		 SET title = $1, summary = $2, category_id = $3, raw_text_vector = $4::vector 
		 WHERE id = $5`,
		title, summary, finalCategoryId, string(embeddingJson), job.BookmarkID,
	)
	if err != nil {
		return fmt.Errorf("failed to save bookmark metadata: %w", err)
	}

	// Insert and associate tags
	for _, tagName := range tags {
		var tagId int
		err = db.Pool.QueryRow(context.Background(),
			`INSERT INTO tags (user_id, name) VALUES ($1, $2) 
			 ON CONFLICT (user_id, name) DO UPDATE SET name = EXCLUDED.name 
			 RETURNING id`,
			job.UserID, tagName,
		).Scan(&tagId)
		if err == nil {
			_, _ = db.Pool.Exec(context.Background(),
				"INSERT INTO bookmark_tags (bookmark_id, tag_id) VALUES ($1, $2) ON CONFLICT DO NOTHING",
				job.BookmarkID, tagId,
			)
		}
	}

	// Encrypt and upload assets
	if job.EncryptionKeyHex != "" {
		assets := []struct {
			Type        string
			Payload     []byte
			ContentType string
		}{
			{Type: "html_dump", Payload: []byte(htmlDump), ContentType: "text/html"},
			{Type: "screenshot", Payload: screenshotBuf, ContentType: "image/png"},
		}

		if len(pdfBuf) > 0 {
			assets = append(assets, struct {
				Type        string
				Payload     []byte
				ContentType string
			}{Type: "pdf", Payload: pdfBuf, ContentType: "application/pdf"})
		}

		for _, asset := range assets {
			encryptedBytes, ivHex, err := crypto.EncryptBuffer(asset.Payload, job.EncryptionKeyHex)
			if err != nil {
				log.Printf("Encryption failed for asset %s: %v\n", asset.Type, err)
				continue
			}

			hasher := sha256.New()
			hasher.Write(asset.Payload)
			checksum := hex.EncodeToString(hasher.Sum(nil))

			storagePath := fmt.Sprintf("%d/%d/%s.enc", job.UserID, job.BookmarkID, asset.Type)

			_, err = s3.UploadAsset(storagePath, encryptedBytes, asset.ContentType)
			if err != nil {
				log.Printf("S3 upload failed for asset %s: %v\n", asset.Type, err)
				continue
			}

			_, err = db.Pool.Exec(context.Background(),
				`INSERT INTO encrypted_assets (bookmark_id, asset_type, storage_path, initialization_vector, sha256_checksum)
				 VALUES ($1, $2, $3, $4, $5)`,
				job.BookmarkID, asset.Type, storagePath, ivHex, checksum,
			)
			if err != nil {
				log.Printf("Failed to log asset meta in DB: %v\n", err)
			}
		}
	}

	return nil
}

// predictCategoryName queries Ollama or falls back to standard keyword mapping
func predictCategoryName(text, title, urlStr string) string {
	ollamaUrl := getOllamaUrl()
	ollamaModel := getOllamaModel()

	prompt := fmt.Sprintf(`You are a categorization assistant. Given the title, URL, and content snippet of a web page, predict a single, broad category name (e.g., "Technology", "Cooking", "Finance", "Science", "Education", "Lifestyle", "Design", "News").
Respond ONLY with the category name (1-2 words). Do not include any punctuation, quotes, or extra text.
Title: %s
URL: %s
Snippet: %s

Category:`, title, urlStr, truncateString(text, 1000))

	category, err := queryOllama(ollamaUrl, ollamaModel, prompt)
	if err == nil && category != "" {
		cleaned := regexp.MustCompile(`[^a-zA-Z0-9\s/]`).ReplaceAllString(category, "")
		cleaned = strings.TrimSpace(cleaned)
		if len(cleaned) > 2 && len(cleaned) < 30 {
			return cleaned
		}
	}

	// Fallback rule-based matching
	lowerText := strings.ToLower(title + " " + urlStr + " " + truncateString(text, 1000))
	if strings.Contains(lowerText, "github") || strings.Contains(lowerText, "code") || strings.Contains(lowerText, "npm") || strings.Contains(lowerText, "developer") || strings.Contains(lowerText, "react") || strings.Contains(lowerText, "api") {
		return "Development"
	}
	if strings.Contains(lowerText, "recipe") || strings.Contains(lowerText, "cook") || strings.Contains(lowerText, "food") || strings.Contains(lowerText, "kitchen") || strings.Contains(lowerText, "bake") {
		return "Cooking"
	}
	if strings.Contains(lowerText, "stock") || strings.Contains(lowerText, "crypto") || strings.Contains(lowerText, "finance") || strings.Contains(lowerText, "investing") || strings.Contains(lowerText, "money") {
		return "Finance"
	}
	if strings.Contains(lowerText, "design") || strings.Contains(lowerText, "css") || strings.Contains(lowerText, "ui/ux") || strings.Contains(lowerText, "vector") || strings.Contains(lowerText, "color") {
		return "Design"
	}
	if strings.Contains(lowerText, "science") || strings.Contains(lowerText, "physics") || strings.Contains(lowerText, "research") || strings.Contains(lowerText, "nature") || strings.Contains(lowerText, "space") {
		return "Science"
	}

	return "General"
}

// generateSummary summarizes text via Ollama or parses first sentences
func generateSummary(text, title string) string {
	ollamaUrl := getOllamaUrl()
	ollamaModel := getOllamaModel()

	prompt := fmt.Sprintf(`You are a summarization assistant. Read this web page text and write a clean, 2-sentence executive summary.
Title: %s
Content snippet: %s

Summary:`, title, truncateString(text, 3000))

	summary, err := queryOllama(ollamaUrl, ollamaModel, prompt)
	if err == nil && summary != "" {
		return strings.TrimSpace(summary)
	}

	// Fallback local NLP sentence segmenter
	cleanSnippet := regexp.MustCompile(`\s+`).ReplaceAllString(text, " ")
	cleanSnippet = strings.TrimSpace(cleanSnippet)
	sentences := regexp.MustCompile(`(?<=[.!?])\s+`).Split(cleanSnippet, -1)
	
	var summarySentences []string
	summaryLength := 0
	for _, sentence := range sentences {
		if len(sentence) > 15 {
			summarySentences = append(summarySentences, sentence)
			summaryLength += len(sentence)
			if len(summarySentences) >= 2 || summaryLength > 200 {
				break
			}
		}
	}

	if len(summarySentences) > 0 {
		return strings.Join(summarySentences, " ")
	}

	return fmt.Sprintf("Archived link to %s.", title)
}

// generateTags extracts tags via Ollama or word frequency
func generateTags(text, title string) []string {
	ollamaUrl := getOllamaUrl()
	ollamaModel := getOllamaModel()

	prompt := fmt.Sprintf(`You are a taxonomy expert. Output up to 5 precise keywords or tags for this web page.
Respond ONLY with a comma-separated list of lowercase tags. Example: react, javascript, frontend, hooks.
Title: %s
Content snippet: %s

Tags:`, title, truncateString(text, 1500))

	tagsRes, err := queryOllama(ollamaUrl, ollamaModel, prompt)
	if err == nil && tagsRes != "" {
		var tags []string
		for _, part := range strings.Split(tagsRes, ",") {
			cleaned := regexp.MustCompile(`[^a-zA-Z0-9-]`).ReplaceAllString(part, "")
			cleaned = strings.TrimSpace(strings.ToLower(cleaned))
			if len(cleaned) > 1 && len(cleaned) < 20 {
				tags = append(tags, cleaned)
			}
		}
		if len(tags) > 0 {
			return tags
		}
	}

	// Fallback word frequency counter
	blacklist := map[string]bool{
		"the": true, "and": true, "for": true, "with": true, "this": true, "that": true,
		"your": true, "from": true, "have": true, "were": true, "about": true, "should": true,
		"would": true, "could": true, "their": true, "there": true, "these": true, "those": true,
		"home": true, "page": true, "site": true, "website": true, "login": true, "signup": true,
	}

	cleanStr := regexp.MustCompile(`[^a-z0-9\s-]`).ReplaceAllString(strings.ToLower(title+" "+truncateString(text, 1000)), "")
	words := strings.Fields(cleanStr)

	freq := make(map[string]int)
	for _, word := range words {
		if len(word) > 3 && !blacklist[word] {
			freq[word]++
		}
	}

	// Grab top 4
	var tags []string
	for i := 0; i < 4; i++ {
		bestWord := ""
		bestCount := -1
		for w, c := range freq {
			if c > bestCount {
				bestCount = c
				bestWord = w
			}
		}
		if bestWord != "" {
			tags = append(tags, bestWord)
			delete(freq, bestWord)
		}
	}

	if len(tags) == 0 {
		return []string{"web", "bookmark"}
	}
	return tags
}

// getOllamaEmbedding calls the Ollama embeddings API
func getOllamaEmbedding(text, title string) ([]float32, error) {
	ollamaUrl := getOllamaUrl()
	embedModel := os.Getenv("OLLAMA_EMBED_MODEL")
	if embedModel == "" {
		embedModel = "all-minilm"
	}

	prompt := title + "\n" + truncateString(text, 1500)

	payload := map[string]interface{}{
		"model":  embedModel,
		"prompt": prompt,
	}

	bodyBytes, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", ollamaUrl+"/api/embeddings", bytes.NewBuffer(bodyBytes))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("status error: %d", resp.StatusCode)
	}

	var res EmbedResponse
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return nil, err
	}

	return res.Embedding, nil
}

// Helper to check Ollama status
func queryOllama(ollamaUrl, model, prompt string) (string, error) {
	payload := map[string]interface{}{
		"model":  model,
		"prompt": prompt,
		"stream": false,
	}

	bodyBytes, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", ollamaUrl+"/api/generate", bytes.NewBuffer(bodyBytes))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("status error: %d", resp.StatusCode)
	}

	var res GenerateResponse
	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
		return "", err
	}

	return res.Response, nil
}

func getOllamaUrl() string {
	urlStr := os.Getenv("OLLAMA_URL")
	if urlStr == "" {
		urlStr = "http://localhost:11434"
	}
	return strings.TrimSuffix(urlStr, "/")
}

func getOllamaModel() string {
	m := os.Getenv("OLLAMA_MODEL")
	if m == "" {
		m = "llama3"
	}
	return m
}

func truncateString(s string, max int) string {
	if len(s) > max {
		return s[:max]
	}
	return s
}
